### PR TITLE
Add support for discovery of privileged script execution

### PR DESCRIPTION
**Pull request by hawkinsw**

### CODE REVIEW

**Comment by hawkinsw** [comment]
Should you want to test this, you can change the `package.json` file in `tools/browsertime/` like so:

```
diff --git a/tools/browsertime/package.json b/tools/browsertime/package.json
--- a/tools/browsertime/package.json
+++ b/tools/browsertime/package.json
@@ -5,7 +5,7 @@
   "license": "MPL-2.0",
   "dependencies": {},
   "devDependencies": {
-    "browsertime": "https://github.com/mozilla/browsertime/tarball/0e9ed0a4c65910c266a45db5f303274f0ce1db0d"
+    "browsertime": "hawkinsw/browsertime#discover_privileged_scripts"
   },
   "notes(private)": "We don't want to publish to npm, so this is marked as private",
   "private": true
```

You will have to (re)run `mach browsertime --setup` after that!

**Comment by hawkinsw** [comment]
To test, you can run this command:

`mach browsertime http://www.google.com -n 1 --script tools/browsertime/node_modules/browsertime/browserscripts/browser/appConstants.js`

In the output file, you will see the contents of the AppConstants object.

**Comment by hawkinsw** [comment]
Hey @ncalexan -- ready for a review!

**Comment by ncalexan** [comment]
> Should you want to test this, you can change the `package.json` file in `tools/browsertime/` like so:

You can also override the browsertime repo with `env BROWSERTIME=/path/to/git/browsertime ./mach browsertime -- ...`.

File: browserscripts/browser/collectStats.js
Comment (ncalexan) [inline browserscripts/browser/collectStats.js]: This look oddly formatted.  Convert tabs to spaces?

File: browserscripts/browser/collectStats.js
Comment (ncalexan) [inline browserscripts/browser/collectStats.js]: How is this different than `return JSON.parse(stats)`?

File: browserscripts/browser/collectStats.js
Comment (ncalexan) [inline browserscripts/browser/collectStats.js]: Everything is async here already, so there's no need to `await` and `then`.  I expect:
```
const stats = await ChromeUtils.collectPerfStats();
return JSON.parse(stats);
```
to be fine.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js]: This should be used only in test code.  Just a regular `if(...) { throw ... }` should be good.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js:388]: All of these whitespace changes will make this harder to upstream, and harder to merge upstream into our code (which I am doing right now).

Are they generated by your IDE?  Try to figure out how to stop that if so.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js:397]: Same with quoting changes -- don't touch the line if you can help it.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js]: Same thing with not using `assert` -- `throw` is better here.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js]: The style in browsertime is to use `/* */` for JS doc (-ish) things, and `//` for inline comments.  AFAICT.

File: browserscripts/prepostscripts/startPerfStats.js
Comment (ncalexan) [inline browserscripts/prepostscripts/startPerfStats.js]: nit: again, formatting.  Tabs?

File: browserscripts/prepostscripts/startPerfStats.js
Comment (ncalexan) [inline browserscripts/prepostscripts/startPerfStats.js]: Can you fold this "start" script into the module definition for the privileged script, so that we have one module with `start`, `function`, `stop` functions?  Then all the relevant definitions are in one place, and it's probably more clear that the start scripts won't run without the function being run, etc.

File: browserscripts/prepostscripts/appConstants.js
Comment (ncalexan) [inline browserscripts/prepostscripts/appConstants.js]: This is a useful example but probably shouldn't land directly.

File: browserscripts/browser/appConstants.js
Comment (ncalexan) [inline browserscripts/browser/appConstants.js]: Why are we stringifying?  Can't we pass back some (potentially sanitized) object?

File: browserscripts/browser/appConstants.js
Comment (ncalexan) [inline browserscripts/browser/appConstants.js]: nit: `appConstants:`.  Help future us grep for this!

File: browserscripts/browser/collectStats.js
Comment (ncalexan) [inline browserscripts/browser/collectStats.js]: nit: call the file `perfStats` (or `privilegedPerfStats`), for greppability and to follow suite with the existing measurement files, which are called like what they return (`userAgent.js`), not what they are doing (`collectUserAgent.js`).

File: lib/core/engine/iteration.js
Comment (ncalexan) [inline lib/core/engine/iteration.js]: If you really need to coerce to boolean, just `rp = !!so.rp`.

File: lib/core/engine/iteration.js
Comment (ncalexan) [inline lib/core/engine/iteration.js]: What is `typeof undefined`?  I.e., does this really work for Chrome, where this will be `undefined`?

File: lib/core/engine/iteration.js
Comment (ncalexan) [inline lib/core/engine/iteration.js]: All these `!== null` aren't particularly idiomatic.  Just `if (script) { ... }`.  If people give wacky inputs, they can deal with wacky outputs.

File: lib/core/engine/iteration.js
Comment (ncalexan) [inline lib/core/engine/iteration.js]: I started to say: "Explain to me why you need to manage contexts yourself here.  Why can't you use the existing API for privileged execution (https://github.com/mozilla/browsertime/blob/9d7e3715744a065fd69afd0681ee1b192f119876/lib/core/seleniumRunner.js#L295), etc."

But in fact all of this pre/post stuff is already possible, i.e.: https://github.com/bdekoz/browsertime/blob/privileged-telemetry/vendor-scripts/clear-persistent-data-stop.js.

So there are two issues: first, what is the goal of these pre/post changes?  Second, the context is only relevant on the target device, and pre/post scripts run on the host machine.  So I don't quite understand how that's going to work.

Can you separate the pre/post stuff into a part 2 commit that builds on the main privileged measurement commit?  That will be the simplest way for us to isolate the needed pieces.

**Review by ncalexan — CHANGES_REQUESTED** [review]
OK, I've given this a first pass.  Generally:

1. the core measurement functionality looks great.  Well done!
2. can you nail down exactly what the window/worker thing you were seeing was, and how this patch addresses it?  I can't see it.
3. I'd like the pre/post script in a second commit, 'cuz I don't think we're on the same page about the relevant problem to solve.
4. there are many nits, most of which are unimportant.  But not introducing cosmetic changes _is_ important: don't churn unless there's value to churn.  Every line changed is a line that doesn't merge trivially.

Way to push through to a solution!  Excited to see round two.

File: lib/core/seleniumRunner.js
Comment (hawkinsw) [reply lib/core/seleniumRunner.js:388]: They are actually generated by *your* configuration -- I just ran it through eslint using the configuration that the browsertime repo specifies :-) I was absolutely not trying to change anything. I was just trying to follow the rules.

File: lib/core/engine/iteration.js
Comment (hawkinsw) [reply lib/core/engine/iteration.js]: According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof), `typeof undefined` is undefined. I confirmed this is implemented that way on both Chrome and Fx. Are we okay with that?

File: browserscripts/browser/appConstants.js
Comment (hawkinsw) [reply browserscripts/browser/appConstants.js]: Just to clarify, even though the object that we import is from `AppConstants.jsm`, we want to name the object that we retrieve `appConstants`?

Comment (hawkinsw) [reply browserscripts/browser/appConstants.js]: Update: I'm an idiot. Changing the case of that variable will *not* work. The 

`const { AppConstants } = ...`

destructures an object returned by the import statement and assigns the field from the returned object named `AppConstant` to our variable `AppConstant`. If those variable names differ, then there will not be a match and the variable will be null. 

TL;DR: We have to keep it named `AppConstants`.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [reply lib/core/seleniumRunner.js:388]: > They are actually generated by _your_ configuration -- I just ran it through eslint using the configuration that the browsertime repo specifies :-) I was absolutely not trying to change anything. I was just trying to follow the rules.

Well, _mine_ in the sense of "whatever was in the browsertime repo".  It's a good idea to run things through linters, etc before landing -- but upstream/downstream trumps whitespace churn.

Is it possible to apply ESlint only to changed lines?  That's best possible here.

File: lib/core/engine/iteration.js
Comment (ncalexan) [reply lib/core/engine/iteration.js]: Yep!

File: browserscripts/browser/appConstants.js
Comment (ncalexan) [reply browserscripts/browser/appConstants.js]: There are two different things going on here.  One is destructuring: that name must be `AppConstants`.  The other is what we expose, which in your patch is `{constants: ...}`.  Make that `{appConstants: ...}` or `{AppConstants: ...}` so that it's clear what set of constants these are (or, somewhat clear, at least).  `appConstants` is more in keeping with browsertime, which uses `userAgent`, etc.

File: lib/core/engine/iteration.js
Comment (hawkinsw) [reply lib/core/engine/iteration.js]: is 

`if (x === null) {`

idiomatic or should that be 

`if (x) {`

?

File: browserscripts/browser/collectStats.js
Comment (hawkinsw) [reply browserscripts/browser/collectStats.js]: Since we are not committing this file in this revision, I'm going to simply mark this as resolved.

File: lib/core/engine/iteration.js
Comment (hawkinsw) [reply lib/core/engine/iteration.js]: > I started to say: "Explain to me why you need to manage contexts yourself here. Why can't you use the existing API for privileged execution (
> 
> https://github.com/mozilla/browsertime/blob/9d7e3715744a065fd69afd0681ee1b192f119876/lib/core/seleniumRunner.js#L295
> ), etc."
> 
> But in fact all of this pre/post stuff is already possible, i.e.: https://github.com/bdekoz/browsertime/blob/privileged-telemetry/vendor-scripts/clear-persistent-data-stop.js.
> 
> So there are two issues: first, what is the goal of these pre/post changes? Second, the context is only relevant on the target device, and pre/post scripts run on the host machine. So I don't quite understand how that's going to work.

This is technically true. *But*, the pre/post scripts get a `context` that can be indirected to execute a script on the target browser. In the example(s) that I posted, that's exactly how I set the collection mask:

```
let run = () => {
  /*
   * Tell the engine that we want to know about all the possible
   * performance stats that are available!
   */
  ChromeUtils.setPerfStatsCollectionMask(255);
};
await context.selenium.driver.executeScript(run).then(value => {});
```

For that to work, the target browser would have to be "primed" with the privileges required. That, at least, was the way I thought about it. 

I will separate out this change into a subsequent PR against the other changes in this PR once this PR lands ... wow, that's a mouthful! That way, as you said, we can work together to think about the mechanics of if/how this works!

Is that okay?

> 
> Can you separate the pre/post stuff into a part 2 commit that builds on the main privileged measurement commit? That will be the simplest way for us to isolate the needed pieces.

**Comment by hawkinsw** [comment]
Okay @ncalexan I think that the version up now should address everything that you requested. Thank you for the detailed feedback (obviously) -- I hope that this one gets us closer to the goal!

**Comment by hawkinsw** [comment]
Yes! Sorry I misunderstood your original comment. That makes perfect sense.

File: lib/core/engine/iteration.js
Comment (ncalexan) [reply lib/core/engine/iteration.js]: Generally just `if (x) { }`.  Sometimes the difference matters, but generally it doesn't; anytime there's an object equality one has to think about what exactly is being guarded against.  Here, it's just "is this set", so no need for strict object equality.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js]: `requires && requires.privilege`, so that callers don't have to upgrade.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js]: Maybe `log.debug`.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js]: nit: Drop the `//\n` lines (throughout).  The formatting (from ESlint?) is already expanding vertical spacing by a great deal.

File: lib/core/seleniumRunner.js
Comment (ncalexan) [inline lib/core/seleniumRunner.js:552]: nit: fewer vertical lines.

File: lib/support/browserScript.js
Comment (ncalexan) [inline lib/support/browserScript.js]: `ScriptObject` isn't actually a class; consider just `generate a "script object"`.

File: lib/support/browserScript.js
Comment (ncalexan) [inline lib/support/browserScript.js]: uber nit: this all seems too horizontally narrow.  This is <60 chars; I'm sitting here able to view 4k.

File: lib/support/browserScript.js
Comment (ncalexan) [inline lib/support/browserScript.js]: Can you say how the signal is used?  You fall back to an old style script?

**Review by ncalexan — APPROVED** [review]
OK, I think this looks good!  I have some nits but everything looks like I expect.

I'll try to take it for a spin, but that doesn't have to block landing.

File: lib/core/seleniumRunner.js
Comment (hawkinsw) [reply lib/core/seleniumRunner.js]: Good eyes!

File: lib/support/browserScript.js
Comment (hawkinsw) [reply lib/support/browserScript.js]: Show off.

File: lib/support/browserScript.js
Comment (hawkinsw) [reply lib/support/browserScript.js]: Great suggestion!

File: lib/support/browserScript.js
Comment (hawkinsw) [reply lib/support/browserScript.js]: I waffled on this. Thanks for the definitive answer!

File: lib/core/seleniumRunner.js
Comment (hawkinsw) [reply lib/core/seleniumRunner.js:552]: Done!

File: lib/core/seleniumRunner.js
Comment (hawkinsw) [reply lib/core/seleniumRunner.js]: Just carryover of personal style, that's all. I can blame eslint for many ills -- this is not one of them! :-)

File: lib/core/seleniumRunner.js
Comment (hawkinsw) [reply lib/core/seleniumRunner.js]: Loquacious.

**Comment by ncalexan** [comment]
I tested this locally, and it's not quite right.  Uber-nit: bump the log-levels so that the
```
[2019-08-26 14:40:42] INFO: [browsertime] timeToFirstInteractive is an old-style script object.
```
lines are visible with `-vv` or `-vvv` only.  (Debug?  Trace?)  Remove the
```
[2019-08-26 14:40:49] DEBUG: [browsertime] Running script loadEventEnd without privilege.
```
lines entirely (or bump them to trace).  The `with privilege` lines are helpful.

This doesn't work in recent Chrome (76):
```
[2019-08-26 14:40:49] ERROR: [browsertime] Failed to execute user script: BrowserError: Only Firefox browsers can run privileged JavaScript: chrome
...
[2019-08-26 14:40:49] DEBUG: [browsertime] Telling browser to quit.
[2019-08-26 14:40:49] DEBUG: [browsertime] Stopping Browsertime
[2019-08-26 14:40:49] DEBUG: [browsertime] Stopped Browsertime
[2019-08-26 14:40:49] ERROR: [browsertime] Error running browsertime Error: Unhandled value type undefined found when adding data
    at addRecursive (/Users/nalexander/Mozilla/browsertime/lib/support/statistics.js:106:17)
    at Object.keys.forEach.key (/Users/nalexander/Mozilla/browsertime/lib/support/statistics.js:87:15)
    at Array.forEach (<anonymous>)
    at addRecursive (/Users/nalexander/Mozilla/browsertime/lib/support/statistics.js:85:32)
    at Statistics.addDeep (/Users/nalexander/Mozilla/browsertime/lib/support/statistics.js:112:5)
    at Collector.perIteration (/Users/nalexander/Mozilla/browsertime/lib/core/engine/collector.js:125:20)
    at Engine.runByScript (/Users/nalexander/Mozilla/browsertime/lib/core/engine/index.js:160:25)
    at process.internalTickCallback (internal/process/next_tick.js:77:7)
```
But on Firefox, everything looks grand.

**Comment by hawkinsw** [comment]
> I tested this locally, and it's not quite right. Uber-nit: bump the log-levels so that the
> 
> ```
> [2019-08-26 14:40:42] INFO: [browsertime] timeToFirstInteractive is an old-style script object.
> ```
> 
> lines are visible with `-vv` or `-vvv` only. (Debug? Trace?) Remove the
> 
> ```
> [2019-08-26 14:40:49] DEBUG: [browsertime] Running script loadEventEnd without privilege.
> ```
> 
> lines entirely (or bump them to trace). The `with privilege` lines are helpful.
> 
> This doesn't work in recent Chrome (76):
> 
> ```
> [2019-08-26 14:40:49] ERROR: [browsertime] Failed to execute user script: BrowserError: Only Firefox browsers can run privileged JavaScript: chrome
> ...
> [2019-08-26 14:40:49] DEBUG: [browsertime] Telling browser to quit.
> [2019-08-26 14:40:49] DEBUG: [browsertime] Stopping Browsertime
> [2019-08-26 14:40:49] DEBUG: [browsertime] Stopped Browsertime
> [2019-08-26 14:40:49] ERROR: [browsertime] Error running browsertime Error: Unhandled value type undefined found when adding data
>     at addRecursive (/Users/nalexander/Mozilla/browsertime/lib/support/statistics.js:106:17)
>     at Object.keys.forEach.key (/Users/nalexander/Mozilla/browsertime/lib/support/statistics.js:87:15)
>     at Array.forEach (<anonymous>)
>     at addRecursive (/Users/nalexander/Mozilla/browsertime/lib/support/statistics.js:85:32)
>     at Statistics.addDeep (/Users/nalexander/Mozilla/browsertime/lib/support/statistics.js:112:5)
>     at Collector.perIteration (/Users/nalexander/Mozilla/browsertime/lib/core/engine/collector.js:125:20)
>     at Engine.runByScript (/Users/nalexander/Mozilla/browsertime/lib/core/engine/index.js:160:25)
>     at process.internalTickCallback (internal/process/next_tick.js:77:7)
> ```
> 
> But on Firefox, everything looks grand.

Check the latest revision of the commit. I think that we are good-to-go! :-)

**Comment by ncalexan** [comment]
OK, this is basically ready.

1. Bump up the commit message, or add some other documentation (in `README.md`).  Explain the trade-offs.

2. Land with `appConstants.js`, so there's an example in the tree.

3.  Make these "agree":
```
[2019-08-28 11:44:02] VERBOSE: [browsertime] Executing script visualElements
[2019-08-28 11:44:02] DEBUG: [browsertime] Running script appConstants with privilege.
```
So debug -> verbose, running -> executing.

4. Turn down the level for `PrivilegeError`.  That should be info or warning, I think.
```
[2019-08-28 11:45:34] ERROR: [browsertime] Failed to execute user script: BrowserError: Only Firefox browsers can run privileged JavaScript: chrome
```
And be clear that we're just skipping and moving on; there's no user action required here.  Maybe even make this debug; it's just not something "regular" consumers think about.

I will try to make sure you have merge access so you can polish it up and bombs away.

**Comment by ncalexan** [comment]
Merged with small nits addressed: https://github.com/mozilla/browsertime/commit/fe0a34d2044e0f81a36b4ecec2ecf140162428ec.  Great work, Will!

**Pull request closed without merging**
